### PR TITLE
Feature/ui data modeling post3.0

### PR DESCRIPTION
**Pull request by ajainarayanan**
Finishing up what was started here (https://github.com/caskdata/cdap/pull/1498)

The last update from me in PR#1498 (https://github.com/caskdata/cdap/pull/1498#issuecomment-82712166) is still valid except for 1st TODO.

In addition: 
- We need to modify ```BuildUrl``` function in datasource.js to parse params for metrics (tags based url.)
   The reasoning behind this is that the metrics url will have the same query string (tags) and we cannot pass a json map with same key as property multiple times.

```
  { 
    tag: 'apps:PurchaseHistory',
    tag: 'workflows:PurchaseHistoryWorkflow',
    tag: 'metric:Something'
  } 
```
the snippet is not a valid json. We need to find way to pass an array of objects to construct query string.( For further details check https://docs.angularjs.org/api/ngResource/service/$resource USAGE section)

### CODE REVIEW

File: cdap-ui/app/features/workflows/controllers/tabs/runs/tabs/log-ctrl.js [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
     if (!$scope.runs.length) {
       return;
     }
-    dataSrc.poll({
-      _cdapNsPath: basePath + '/logs/next?max=50'
-    }, function(res) {
-      $scope.logs = res;
-    });
-
+    params.max = 50;
```
Comment (elfenheart) [inline cdap-ui/app/features/workflows/controllers/tabs/runs/tabs/log-ctrl.js]: Can we include this params.max within line 3-8?

Comment (ajainarayanan) [inline cdap-ui/app/features/workflows/controllers/tabs/runs/tabs/log-ctrl.js]: Fixed.

**Comment by anwar6953** [comment]
In response to what you said about the query tags, can we instead have a field/object that is `tags`, and it simply be a string to string mapping.
For instance:
` { 
    'apps' : 'PurchaseHistory',
    'workflows' : 'PurchaseHistoryWorkflow',
    'metric' : 'Something'
  } `
Then, building the url can take that dictionary, and for each key-value pairing, build a segment of the url like `tag=key:value`.
This avoids having the same string for multiple key-value pairings, right?

File: cdap-ui/app/services/data/datasource.js [hunk 2]
```diff
@@ -199,30 +276,92 @@ angular.module(PKG.name+'.services')
...
+function buildUrl(url, params) {
+  if (!params) return url;
+  var parts = [];
+
+  function forEachSorted(obj, iterator, context) {
+    var keys = Object.keys(params).sort();
+    for (var i = 0; i < keys.length; i++) {
+      iterator.call(context, obj[keys[i]], keys[i]);
+    }
+  return keys;
```
Comment (anwar6953) [inline cdap-ui/app/services/data/datasource.js]: indent is off

**Comment by ajainarayanan** [comment]
@anwar6953 : It is something that is done only in operations. We could have generic method that accepts a js object and constructs the URL for us. Once we have the url with proper query parameters, then we can directly provide the url to ```$resource``` object which will make the request.

The contention is: is it worth abstracting out the operations part because eventually the controller is constructing the url instead of the factory or service (no re-use). Probably we end up operations using raw MyDataSource and everything else using ```$resource```

File: cdap-ui/app/services/data/datasource.js (hunk 2 above)
Comment (ajainarayanan) [inline cdap-ui/app/services/data/datasource.js]: Fixed.

**Comment by anwar6953** [comment]
Would something like this be useful?
https://github.com/caskdata/cdap/pull/2718/files#diff-f5be262a9ebc0b228914ac38fba4796bR164

**Comment by elfenheart** [comment]
:+1: LGTM

**Pull request merged**